llvm/lib/Target/DirectX/DXILRootSignature.cpp

```diff
@@ -0,0 +1,229 @@
+    return reportError(Ctx, "Invalid Root Signature Element: " +
+                                ElementText->getString());
+  }
+
+  llvm_unreachable("Unhandled RootSignatureElementKind enum.");
+}
+
+static bool parse(LLVMContext *Ctx, mcdxbc::RootSignatureDesc &RSD,
+                  MDNode *Node) {
+  bool HasError = false;
+
+  // Loop through the Root Elements of the root signature.
+  for (const auto &Operand : Node->operands()) {
+    MDNode *Element = dyn_cast<MDNode>(Operand);
+    if (Element == nullptr)
+      return reportError(Ctx, "Missing Root Element Metadata Node.");
+
+    HasError = HasError || parseRootSignatureElement(Ctx, RSD, Element);
+  }
+
+  return HasError;
+}
+
+static bool validate(LLVMContext *Ctx, const mcdxbc::RootSignatureDesc &RSD) {
+  if (!dxbc::RootSignatureValidations::isValidRootFlag(RSD.Flags)) {
+    return reportError(Ctx, "Invalid Root Signature flag value");
+  }
+  return false;
+}
+
+static SmallDenseMap<const Function *, mcdxbc::RootSignatureDesc>
+analyzeModule(Module &M) {
+
+  /** Root Signature are specified as following in the metadata:
+
+    !dx.rootsignatures = !{!2} ; list of function/root signature pairs
+    !2 = !{ ptr @main, !3 } ; function, root signature
+    !3 = !{ !4, !5, !6, !7 } ; list of root signature elements
+
+    So for each MDNode inside dx.rootsignatures NamedMDNode
+    (the Root parameter of this function), the parsing process needs
+    to loop through each of its operands and process the function,
+    signature pair.
+ */
+
+  LLVMContext *Ctx = &M.getContext();
+
+  SmallDenseMap<const Function *, mcdxbc::RootSignatureDesc> RSDMap;
+
+  NamedMDNode *RootSignatureNode = M.getNamedMetadata("dx.rootsignatures");
+  if (RootSignatureNode == nullptr)
+    return RSDMap;
+
+  for (const auto &RSDefNode : RootSignatureNode->operands()) {
+    if (RSDefNode->getNumOperands() != 2) {
+      reportError(Ctx, "Invalid format for Root Signature Definition. Pairs "
+                       "of function, root signature expected.");
+      continue;
+    }
+
+    // Function was pruned during compilation.
+    const MDOperand &FunctionPointerMdNode = RSDefNode->getOperand(0);
+    if (FunctionPointerMdNode == nullptr) {
+      reportError(
+          Ctx, "Function associated with Root Signature definition is null.");
+      continue;
+    }
+
+    ValueAsMetadata *VAM =
+        llvm::dyn_cast<ValueAsMetadata>(FunctionPointerMdNode.get());
+    if (VAM == nullptr) {
+      reportError(Ctx, "First element of root signature is not a Value");
+      continue;
+    }
+
+    Function *F = dyn_cast<Function>(VAM->getValue());
+    if (F == nullptr) {
+      reportError(Ctx, "First element of root signature is not a Function");
+      continue;
+    }
+
+    MDNode *RootElementListNode =
+        dyn_cast<MDNode>(RSDefNode->getOperand(1).get());
+
+    if (RootElementListNode == nullptr) {
+      reportError(Ctx, "Missing Root Element List Metadata node.");
+    }
```

Comment:
Should there be a `continue` here? It looks to me like we'll go ahead and use a nullptr later if this happens.